gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/encryption/EncryptionConfiguration.java

```diff
@@ -0,0 +1,162 @@
+            return URI.create(uriWithoutCredentials).getPath().substring(1);
+        }
+
+        return environment.getProperty("management.mongodb.dbname", "gravitee");
+    }
+
+    private Map<String, Map<String, Object>> loadKmsProvidersFromConfig() {
+        Map<String, Map<String, Object>> kmsProviders = new HashMap<>();
+        boolean found = true;
+        int idx = 0;
+
+        while (found) {
+            String type = environment.getProperty(KEY_MANAGEMENT_SYSTEM_PROPERTY + "[" + idx + "].type");
+            found = (type != null);
+            if (found) {
+                switch (type) {
+                    case KEY_MANAGEMENT_SYSTEM_LOCAL_TYPE:
+                        kmsProviders.put(type, loadLocalKmsProviderFromConfig(idx));
+                        break;
+                    default:
+                        log.warn("Unsupported key management system type: {}", type);
+                        break;
+                }
+            }
+            idx++;
+        }
+        if (kmsProviders.isEmpty()) {
+            throw new IllegalStateException("No key management system provider found");
+        }
+        return kmsProviders;
+    }
+
+    private Map<String, Object> loadLocalKmsProviderFromConfig(int idx) {
+        Map<String, Object> localKmsProvider = new HashMap<>();
+        String masterKey = environment.getProperty(KEY_MANAGEMENT_SYSTEM_PROPERTY + "[" + idx + "].local.key");
+        if (masterKey == null) {
+            throw new IllegalStateException("Master key not set");
+        }
+        localKmsProvider.put(KEY_MANAGEMENT_SYSTEM_LOCAL_KEY_PROPERTY, masterKey);
+
+        return localKmsProvider;
+    }
+
+    @Bean
+    @Conditional(EncryptionEnabledCondition.class)
+    public ClientEncryption clientEncryption(MongoFactory mongoFactory) {
+        // Key Management System (KMS) providers.
+        Map<String, Map<String, Object>> kmsProviders = loadKmsProvidersFromConfig();
+
+        ClientEncryptionSettings.Builder builder = ClientEncryptionSettings
+            .builder()
+            // The collection in MongoDB where the Data Encryption Keys (DEKs) will be stored.
+            .keyVaultNamespace(getDatabaseName() + "." + tablePrefix + keyVaultCollectionName)
+            .keyVaultMongoClientSettings(mongoFactory.buildClientSettingsFromProperties(false))
+            .kmsProviders(kmsProviders);
+
+        ClientEncryption clientEncryption = ClientEncryptions.create(builder.build());
+
+        // Initialize the Data Encryption Key if not already existing.
+        initDataEncryptionKey(clientEncryption);
+
+        return clientEncryption;
+    }
+
+    private void initDataEncryptionKey(ClientEncryption clientEncryption) {
+        BsonDocument keyByAltName = clientEncryption.getKeyByAltName(keyAlternativeName);
+        if (keyByAltName == null) {
+            clientEncryption.createDataKey(KEY_MANAGEMENT_SYSTEM_LOCAL_TYPE, new DataKeyOptions().keyAltNames(List.of(keyAlternativeName)));
```

Comment:
For the record, like we talked about, we can add a try/catch clause to catch the unique violation exception in case two instances try to create the DEK at the same time.